appendices/memorymodel.txt

```diff
@@ -788,21 +788,53 @@ NOTE: It is possible for there to be a write between X and Y that overwrites
 a subset of the memory locations, but the remaining memory locations (M~2~)
 will still be visible-to Y.
 
-[[memory-model-scoped-modification-order-coherence]]
-== Scoped Modification Order Coherence
+[[memory-model-acyclicity]]
+== Acyclicity
+
+_Reads-from_ is a relation between operations, where the first operation is a
+write, the second operation is a read, and the second operation reads the
+value written by the first operation.
+_From-reads_ is a relation between operations, where the first operation is a
+read, the second operation is a write, and the first operation reads a value
+written earlier than the second operation in the second operation's scoped
+modification order (or the first operation reads from the initial value, and
+the second operation is any write to the same locations).
+
+Then the implementation must: guarantee that no cycles exist in the union of
+the following relations:
 
-Let A and B be mutually-ordered atomic operations, where A happens-before B,
-and let O be A's scoped modification order.
-Then:
+  * location-ordered
+  * scoped modification order (over all atomic writes)
+  * reads-from
+  * from-reads
 
-  * If A and B are both writes, then A must: be earlier than B in O
-  * If A and B are both reads, then the write that A takes its value from
-    must: be earlier in O than (or the same as) the write that B takes its
-    value from
+NOTE: This is a "consistency" axiom, which informally guarantees that
+sequences of operations can't violate causality.
+
+[[memory-model-scoped-modification-order-coherence]]
+=== Scoped Modification Order Coherence
```

Comment:
nit: I dislike the use of the word "coherence" due to all the reasons we've previously discussed about this term ;). If there's no sane alternative or if this is mirroring some C++ term then it's fine, but I figured it was worth at least mentioning.

Reply:
This is mirroring C++ terminology, each of these rules has a note like: "[Note: This requirement is known as read-read coherence. —end note]"